zebra/kernel_netlink.c

```diff
@@ -1080,6 +1132,288 @@ int netlink_request(struct nlsock *nl, void *req)
+{
+	bth->buf = nl_batch_tx_buf;
+	bth->bufsiz = sizeof(nl_batch_tx_buf);
+	bth->limit = NL_BATCH_SEND_THRESHOLD;
+
+	bth->buf_head = bth->buf;
+	bth->curlen = 0;
+	bth->msgcnt = 0;
+	bth->zns = NULL;
+
+	nl_batch_list_init(&(bth->items));
+}
+
+static void nl_batch_send(struct nl_batch *bth)
+{
+	struct nl_msg_batch_item *item;
+	bool err = false;
+
+	if (bth->curlen == 0 || bth->zns == NULL)
+		return;
+
+	if (IS_ZEBRA_DEBUG_KERNEL)
+		zlog_debug("%s: %s, batch size=%zu, msg cnt=%zu", __func__,
+			   bth->zns->nls.name, bth->curlen, bth->msgcnt);
+
+	if (netlink_send_msg(&(bth->zns->nls), bth->buf, bth->curlen) == -1)
+		err = true;
+
+	if (!err) {
+		if (nl_batch_read_resp(bth) == -1)
+			err = true;
+	}
+
+	frr_each_safe(nl_batch_list, &(bth->items), item) {
+		enum zebra_dplane_result res = ZEBRA_DPLANE_REQUEST_SUCCESS;
+
+		/*
+		 * If either sending or receiving a message batch has ended with
+		 * the error, mark all dplane requests as failed.
+		 */
+		if (item->failure || err)
+			res = ZEBRA_DPLANE_REQUEST_FAILURE;
+
+		if (!item->ignore_res)
+			dplane_ctx_set_status(item->ctx, res);
+
+		nl_batch_list_del(&(bth->items), item);
+		XFREE(MTYPE_NL_BATCH_ITEM, item);
+	}
+
+	nl_batch_reset(bth);
+}
+
+static void nl_batch_add_item(struct nl_batch *bth, int seq,
+			      struct zebra_dplane_ctx *ctx, bool ignore_res)
+{
+	struct nl_msg_batch_item *item =
+		XCALLOC(MTYPE_NL_BATCH_ITEM, sizeof(*item));
+
+	item->seq = seq;
+	item->ctx = ctx;
+	item->ignore_res = ignore_res;
+	item->failure = false;
+
+	nl_batch_list_add_tail(&(bth->items), item);
+}
+
+int netlink_batch_add_msg(struct nl_batch *bth, struct zebra_dplane_ctx *ctx,
+			  ssize_t (*msg_encoder)(struct zebra_dplane_ctx *,
+						 void *, size_t),
+			  bool extra_msg)
+{
+	int seq;
+	ssize_t size;
+	struct nlmsghdr *msgh;
+
+	if (bth->zns != NULL
+	    && bth->zns->ns_id != dplane_ctx_get_ns(ctx)->ns_id)
+		nl_batch_send(bth);
+
+	size = (*msg_encoder)(ctx, bth->buf_head, bth->bufsiz - bth->curlen);
+	if (size < 0)
+		return -1;
+
+	if (size == 0) {
+		nl_batch_send(bth);
+		size = (*msg_encoder)(ctx, bth->buf_head,
+				      bth->bufsiz - bth->curlen);
+		if (size <= 0)
+			return -1;
+	}
+
+	seq = dplane_ctx_get_ns(ctx)->nls.seq;
+	if (extra_msg)
+		seq++;
+
+	msgh = (struct nlmsghdr *)bth->buf_head;
+	msgh->nlmsg_seq = seq;
+	msgh->nlmsg_pid = dplane_ctx_get_ns(ctx)->nls.snl.nl_pid;
+
+	nl_batch_add_item(bth, seq, ctx, extra_msg);
+
+	bth->zns = dplane_ctx_get_ns(ctx);
+	bth->buf_head = ((char *)bth->buf_head) + size;
+	bth->curlen += size;
+	bth->msgcnt++;
+
+	if (bth->curlen > bth->limit)
+		nl_batch_send(bth);
+
+	return 0;
+}
+
+static int nl_put_msg(struct nl_batch *bth, struct zebra_dplane_ctx *ctx)
+{
+	if (dplane_ctx_is_skip_kernel(ctx))
+		return 1;
+
+	switch (dplane_ctx_get_op(ctx)) {
+
+	case DPLANE_OP_ROUTE_INSTALL:
+	case DPLANE_OP_ROUTE_UPDATE:
+	case DPLANE_OP_ROUTE_DELETE:
+		return netlink_put_route_update_msg(bth, ctx);
```

Comment:
These were a little convoluted  - we add new apis that get called here, and then end up ... 50 lines above, with different encoder callbacks. why not just have the callbacks available, and skip the 'put_xxx_msg' layer here - is that providing some value?

Reply:
Again, there is the problem with context objects that require two Netlink messages. I know, this looks confusing, but I couldn't came up with more sensible solution.

Reply:
ok, yes, I get what you're saying